subsys/bluetooth/mesh/pb_gatt_srv.c

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ * Copyright (c) 2021 Lingao Meng
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+
+#include <net/buf.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/mesh.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_PROXY)
+#define LOG_MODULE_NAME bt_mesh_pb_gatt_srv
+#include "common/log.h"
+
+#include "mesh.h"
+#include "adv.h"
+#include "net.h"
+#include "rpl.h"
+#include "transport.h"
+#include "host/ecc.h"
+#include "prov.h"
+#include "beacon.h"
+#include "foundation.h"
+#include "access.h"
+#include "proxy.h"
+#include "proxy_msg.h"
+#include "pb_gatt_srv.h"
+
+#define CLIENT_BUF_SIZE 66
+
+#if defined(CONFIG_BT_MESH_DEBUG_USE_ID_ADDR)
+#define ADV_OPT_USE_IDENTITY BT_LE_ADV_OPT_USE_IDENTITY
+#else
+#define ADV_OPT_USE_IDENTITY 0
+#endif
+
+#if defined(CONFIG_BT_MESH_PROXY_USE_DEVICE_NAME)
+#define ADV_OPT_USE_NAME BT_LE_ADV_OPT_USE_NAME
+#else
+#define ADV_OPT_USE_NAME 0
+#endif
+
+#define ADV_OPT_PROV                                                           \
+	(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_SCANNABLE |                 \
+	 BT_LE_ADV_OPT_ONE_TIME | ADV_OPT_USE_IDENTITY |                       \
+	 BT_LE_ADV_OPT_USE_NAME)
+
+#define ADV_SLOW_INT                                                           \
+	.interval_min = BT_GAP_ADV_SLOW_INT_MIN,                               \
+	.interval_max = BT_GAP_ADV_SLOW_INT_MAX
+
+#define ADV_FAST_INT                                                           \
+	.interval_min = BT_GAP_ADV_FAST_INT_MIN_2,                             \
+	.interval_max = BT_GAP_ADV_FAST_INT_MAX_2
+
+static bool prov_fast_adv;
+
+static int gatt_send(struct bt_conn *conn,
+		     const void *data, uint16_t len,
+		     bt_gatt_complete_func_t end, void *user_data);
+
+static uint8_t __noinit client_buf_data[CLIENT_BUF_SIZE];
+static struct bt_mesh_proxy_role cli = {
+	.cb.send = gatt_send,
+	.buf = {
+		.__buf = client_buf_data,
+		.data  = client_buf_data,
+		.size  = CLIENT_BUF_SIZE,
+		.len   = CLIENT_BUF_SIZE,
+	},
```

Comment:
Now that this buffer is being instantiated in more places, it seems like it'd be better to just put it in proxy_msg.c, so we don't need to allocate 66 + 30 * conn_count bytes for this. The buffer operation is happening in that file anyway, and there's no scenario where we would run more than `conn_count` number of parallel buffers, regardless of roles.

With this change, I think it'd make sense to move the common part of the gatt connect/disconnect callbacks to that file as well, as they're doing the same thing in both files. If that sits in proxy_msg.c, we only need to handle the "add proxy service" part of the pb_gatt callback.

Reply:
but, how define buf in proxy_msg, there are different length between pbgatt and proxy.

Reply:
Good idea for me👍

Reply:
> but, how define buf in proxy_msg, there are different length between pbgatt and proxy.

perhaps this needs to be `#ifdef` wrapped? ie if GATT proxy is available, it's 30 * conn_count , else it's 66. Could also use a max

Reply:
I think it should be. It seems unrealistic to completely eliminate config in this file, but it still makes sense to move proxy_complete_msg to the corresponding role.

Reply:
One detail needs to be clarified. I think for pb-gatt, only one conn is needed. Therefore, max may be used for the first buf, and 30 for the latter?

Reply:
Before putting conn_cb in the rom section, it is really friendly to our current approach, the structure is very flexible (at least I think it is), and it also saves some RAM resources.

Reply:
In trying this scheme, I found that proxy_msg does not have the ability to obtain client (defined in proxy_srv.c) and cli (defined in pb_gatt_srv.c).

Reply:
I suppose parts of the role structure would have to move as well?

Reply:
how to do?

Reply:
If we understand correctly, do you want to define the role structure in the proxy_msg.c, referenced by proxy_srv.c or pb_gatt_srv.c?

Reply:
``` C
static struct bt_mesh_proxy_client {
	struct bt_mesh_proxy_role *cli;
	uint16_t filter[CONFIG_BT_MESH_PROXY_FILTER_SIZE];
	enum __packed {
		NONE,
		WHITELIST,
		BLACKLIST,

```

this ?

Reply:
Seems like a sensible solution to me, but your call

Reply:
Done.